hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/DiskBasedMap.java

```diff
@@ -0,0 +1,307 @@
+
+    public byte[] getKey() {
+      return key;
+    }
+
+    public byte[] getValue() {
+      return value;
+    }
+
+    public long getTimestamp() {
+      return timestamp;
+    }
+  }
+
+  protected DiskBasedMap(Schema schema, String payloadClazz, Optional<String> baseFilePath) throws IOException {
+    this.inMemoryMetadataOfSpilledData = new HashMap<>();
+
+    if(!baseFilePath.isPresent()) {
+      baseFilePath = Optional.of(DEFAULT_BASE_FILE_PATH);
+    }
+    this.filePath = baseFilePath.get() + UUID.randomUUID().toString();
+    File writeOnlyFileHandle = new File(filePath);
+    initFile(writeOnlyFileHandle);
+
+    this.writeOnlyFileHandle = new SizeAwareDataOutputStream(new FileOutputStream(writeOnlyFileHandle, true));
+    this.filePosition = new AtomicLong(0L);
+    this.schema = schema;
+    this.payloadClazz = payloadClazz;
+  }
+
+  private void initFile(File writeOnlyFileHandle) throws IOException {
+    // delete the file if it exists
+    if(writeOnlyFileHandle.exists()) {
+      writeOnlyFileHandle.delete();
+    }
+    writeOnlyFileHandle.createNewFile();
+    // Open file in readFromDisk-only mode
+    readOnlyFileHandle = new RandomAccessFile(filePath, "r");
+    readOnlyFileHandle.seek(0);
+    // Make sure file is deleted when JVM exits
+    writeOnlyFileHandle.deleteOnExit();
+  }
+
+  /**
+   * Custom iterator to iterate over values written to disk
+   * @return
+   */
+  public Iterator<R> iterator() {
+    return new LazyFileIterable(readOnlyFileHandle,
+        inMemoryMetadataOfSpilledData, schema, payloadClazz).iterator();
+  }
+
+  /**
+   * Number of bytes spilled to disk
+   * @return
+   */
+  public long sizeOfFileOnDiskInBytes() {
+    return filePosition.get();
+  }
+
+  @Override
+  public int size() {
+    return inMemoryMetadataOfSpilledData.size();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return inMemoryMetadataOfSpilledData.isEmpty();
+  }
+
+  @Override
+  public boolean containsKey(Object key) {
+    return inMemoryMetadataOfSpilledData.containsKey(key);
+  }
+
+  @Override
+  public boolean containsValue(Object value) {
+    throw new HoodieNotSupportedException("unable to compare values in map");
+  }
+
+  @Override
+  public R get(Object key) {
+    ValueMetadata entry = inMemoryMetadataOfSpilledData.get(key);
+    if(entry == null) {
+      return null;
+    }
+    try {
+      return SpillableMapUtils.readFromDisk(readOnlyFileHandle, schema,
+          payloadClazz, entry.getOffsetOfValue(), entry.getSizeOfValue());
+    } catch(IOException e) {
+      throw new HoodieIOException("Unable to readFromDisk Hoodie Record from disk", e);
+    }
+  }
+
+  @Override
+  public R put(T key, R value) {
+    //TODO (na) : check value instanceof HoodieRecordPayload, now assume every payload is HoodieRecord
+    HoodieRecord payload = (HoodieRecord) value;
+    try {
+      byte [] val = HoodieAvroUtils.avroToBytes((GenericRecord) payload.getData().getInsertValue(this.schema).get());
```

Comment:
could this happen outside in the caller of this method, so value remains an opaque byte[] here?

Reply:
The Map's types are <Key, HoodieRecord>

Reply:
@vinothchandar ^